### PR TITLE
Switch to updated decorators plugin in Babylon

### PR DESCRIPTION
**Pull request by existentialism**
Fixes https://github.com/prettier/prettier/issues/2628

This probably needs to be a breaking change as users using the old syntax w/ Babylon will now see errors, primarily:

```js
@foo
export default class {}
```

Now needs to be:

```js
export default @foo class{}
```

### CODE REVIEW

File: tests/decorators/redux.js [hunk 1]
```diff
@@ -1,5 +0,0 @@
-@connect(mapStateToProps, mapDispatchToProps)
-export class MyApp extends React.Component {}
-
-@connect(state => ({ todos: state.todos }))
-export class Home extends React.Component {}
```
Comment (vjeux) [inline tests/decorators/redux.js]: Can you keep all those tests somehow and fix them? They are real piece of code that broke in the past, I want to make sure that we keep them as regression test.

**Comment by vjeux** [comment]
Arg, this is very unfortunate that the two syntaxes are not compatible. I'm not really sure how to do deal with it. It would suck to upgrade prettier and not being able to keep using it without changing the babel transform as well. Especially since the new transform is not even officially released yet.

**Comment by existentialism** [comment]
@vjeux yeah, we can close and keep it on ice until at least https://github.com/babel/babel/pull/6107 lands?

**Comment by azz** [comment]
Then there will be two variants of the spec supported (not sure how typescript plans to address the spec change). 

I feel like decorators are more trouble than they're worth :(

**Comment by bakkot** [comment]
Honestly, this is *always* going to be a problem with trying to support pre-stage-3 proposals. From the committee's perspective, the syntax (and semantics, though that's less relevant) of all proposals before stage 3 is basically totally undecided. Anyone using them should be doing so in full awareness that they might change at any time.

Now, I realize a lot of people have no idea about any of this. Partly, though, that's the fault of tools like prettier, which will happily let you use them as if they're part of JavaScript already. I'm kind of split on what this project's stance should be: from one perspective, it's a shame to not support people who do want to use experimental features; from another, it's going to lead to this sort of issue all the time.

Loathe though I generally am to suggest flags, would it be possible to have a flag which enabled pre-stage-3 features, which was a.) off by default and b.) very clear that it might change the syntax in minor version bumps?

File: tests/decorators/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -1,5 +1,47 @@
...
+  
+}
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+@foo("bar")
+class Foo {}
+
+@abc
+class Foo {}
+
+var foo = @dec
```
Comment (azz) [inline tests/decorators/__snapshots__/jsfmt.spec.js.snap]: I think the `class` should be on the same line as the `@dec`.

**Comment by azz** [comment]
> It would suck to upgrade prettier and not being able to keep using it without changing the babel transform as well. Especially since the new transform is not even officially released yet.

It would suck a _bit_, but once https://github.com/babel/babel/pull/6107 is released I imagine it will be easy to code-mod illegal decorator code to the new syntax while upgrading prettier.

I don't feel it's prettier's job to say which proposals you should or shouldn't use. We should however:

1. Make it clear (in release notes, etc) that early proposals are prone to breaking changes.
2. Do work to make sure the migration path in case of breaking changes is as simple as possible.

Summary of options we can take:

1. Support babylon's `"decorators2"` and `"decorators"` simultaneously by way of automatic fallback.
2. Support only `"decorators2"`, This would be a breaking change, but we could ship a codemod with the release.  
3. Support both plugins via a flag to use the old `"decorators"` plugin. 👎

**Comment by vjeux** [comment]
The whole "let's put in the readme that unstable things are going to break" doesn't work. Every single person using mobx is using the old decorator proposal and that's a ton of people. Also typescript uses the old proposal.

I think that the only path forward here is to implement a fallback like we do for the different parsers. I'd love if you could implement it :)

Telling people to upgrade babel is a pretty shitty thing to do, in my experience upgrading babel is not a smooth ride and bunch of things break in the process. I want prettier to remove pain from people, not force them to upgrade one of the core dependencies and run a coremod if they want to keep using it.

**Comment by azz** [comment]
Yeah - the more I think about it the more it makes sense to go with option 1. I'll implement it on this PR.

**Comment by azz** [comment]
Implemented the fallback.

The the following re-print safe?

```js
// Before
@foo
export default class {}

// After
export default @foo
class {}
```

(i.e. will it work with the existing babel transform?)

**Comment by azz** [comment]
@existentialism any thoughts on this one? It's difficult dealing with breaking changes but we'll have to land this in Prettier eventually.

**Comment by existentialism** [comment]
@azz fine with the fallback for now, but uneasy about the combo matrix getting crazy... or worse, needing to also support multiple Babylon versions (if/when decorators2 -> decorators, unsure if this will happen anytime soon, if ever, though).

**Comment by j-f1** [comment]
What’s the status on this PR? Should we merge it now or close it and try again when the Babel PR linked above gets closed?

**Comment by existentialism** [comment]
I think it's fine to close for now, there's quite a bit that's still in flux wrt decorators on the Babel side (ref: https://github.com/babel/babel/pull/7105).

**Pull request closed without merging**